forge-game/src/main/java/forge/game/staticability/StaticAbilityDisableTriggers.java

```diff
@@ -0,0 +1,129 @@
+        // CR 603.2e
+        if (stAb.hasParam("ValidCard") && regtrig.getSpawningAbility() != null) {
+            return false;
+        }
+
+        if (!stAb.matchesValidParam("ValidCard", regtrig.getHostCard())) {
+            return false;
+        }
+
+        if (stAb.hasParam("ValidMode")) {
+            if (!ArrayUtils.contains(stAb.getParam("ValidMode").split(","), regtrig.getMode().toString())) {
+                return false;
+            }
+        }
+
+        if (triggerType.equals(TriggerType.ChangesZone)) {
+            // Cause of the trigger – the card changing zones
+            if (!stAb.matchesValidParam("ValidCause", runParams.get(AbilityKey.Card))) {
+                return false;
+            }
+            if (!stAb.matchesValidParam("Destination", runParams.get(AbilityKey.Destination))) {
+                return false;
+            }
+            if (!stAb.matchesValidParam("Origin", runParams.get(AbilityKey.Origin))) {
+                return false;
+            }
+            if ("Graveyard".equals(runParams.get(AbilityKey.Destination))
+                    && "Battlefield".equals(runParams.get(AbilityKey.Origin))) {
+                // Allow triggered ability of a dying creature that triggers
+                // only when that creature is put into a graveyard from anywhere
+                if ("Card.Self".equals(regtrig.getParam("ValidCard"))
+                        && (!regtrig.hasParam("Origin") || "Any".equals(regtrig.getParam("Origin")))) {
+                    return false;
+                }
+            }
+        } else if (triggerType.equals(TriggerType.ChangesZoneAll)) {
+            final String origin = stAb.getParam("Origin");
+            final String destination = stAb.getParam("Destination");
+            final CardZoneTable table = (CardZoneTable) runParams.get(AbilityKey.Cards);
+
+            // purge all forbidden causes from table
+            CardZoneTable filtered = new CardZoneTable();
+            boolean possiblyDisabled = false;
+            for (Cell<ZoneType, ZoneType, CardCollection> cell : table.cellSet()) {
+                // this currently assumes the table will not contain multiple destinations
+                // however with some effects (e.g. Goblin Welder) that should indeed be the case
+                // once Forge handles that correctly this section needs to account for that
+                // (by doing a closer check of the triggered ability first)
+                CardCollection changers = cell.getValue();
+                if ((origin == null || cell.getRowKey() == ZoneType.valueOf(origin)) &&
+                (destination == null || cell.getColumnKey() == ZoneType.valueOf(destination))) {
+                    changers = CardLists.filter(changers, Predicates.not(CardPredicates.isType(stAb.getParam("ValidCause"))));
```

Comment:
i don't like CardPRedicates.isType there, shouldn't that be more generic with an isValid Predicate?